Move Android Logcat per project settings to AndroidLogcatProjectSettings class

todi1856 opened the pull request:
This PR refactors how we save settings, I moved all serializable settings from AndroidLogcatConsoleWindow to AndroidLogcatProjectSettings, thus moving into direction where we don't need to open UI window for logcat to function.

In regards to logcat project settings, this is a breaking change, since tag settings & memory window settings are serialized differently now

**AndroidLogcatConsoelWindow.cs diff is big, please expand diff to see the changes**

Here's a list of done things:
* Merged all shared functionality to AndroidLogcatRuntimeBase from AndroidLogcatRuntime & AndroidLogcatTestRuntime, thus giving the better testing of runtime.
* Move AndroidLogcatMemoryViewer settings to project settings
* Put PackageInformation in its own file
* Don't use EditorStyles directly, saw instances where they were not initialized.
* Move from AndroidLogcatConsoleWindow ->AndroidLogcatProjectSettings:
  - SelectedPriority
  - SelectedPackage
  - SelectedDevice
* Moved TagInformation to it's own file, refactored TagControl
* Moved AndroidLogcatFakeDevice to its own file
* Remove AndroidLogcatConsoleWindow LoadStates, SaveStates, it's no longer needed
* New tests:
   - Check if we correctly save logcat project settings for domain reloads or window closure
   - Check if we save tags correctly
   - Check if we save selected package correctly
* Columns settings will persist between closing/opening logcat window



Android Project Settings previously looked like this:
```
{
    "m_SelectedDeviceId": "711KPQJ0939020",
    "m_SelectedPackage": {
        "deviceId": "",
        "name": "",
        "processId": 0,
        "exited": false
    },
    "m_SelectedPriority": 0,
    "m_PackagesForSerialization": [
        {
            "deviceId": "711KPQJ0939020",
            "name": " fore   T",
            "processId": 33,
            "exited": true
        }
    ],
    "m_TagControl": {
        "m_TagNames": [
            "Filter by all listed tags",
            "No Filter",
            "",
            "Tag Control...",
            "",
            "Unity",
            "CRASH",
            "Finsky"
        ],
        "m_SelectedTags": [
            false,
            false,
            false,
            false,
            false,
            false,
            false,
            true
        ]
    },
    "m_MemoryViewerJson": "{\"m_MemoryWindowHeight\":300.0,\"m_MemoryWindowWidth\":170.0,\"m_MemoryTypeEnabled\":[true,true,true,true,true,true,true,true,true,true],\"m_MemoryGroup\":1,\"m_MemoryViewerState\":1}"
}
```

now they will look like this;
```
{
    "m_SelectedDeviceId": "711KPQJ0939020",
    "m_SelectedPackage": {
        "deviceId": "711KPQJ0939020",
        "name": "com.Company.AndroidLogcatSample",
        "processId": 2348,
        "exited": false
    },
    "m_SelectedPriority": 0,
    "m_KnownPackagesForSerialization": [
        {
            "deviceId": "711KPQJ0939020",
            "name": "com.google.android.apps.nexuslauncher",
            "processId": 2743,
            "exited": false
        },
        {
            "deviceId": "711KPQJ0939020",
            "name": "com.Company.AndroidLogcatSample",
            "processId": 2348,
            "exited": false
        },
        {
            "deviceId": "192.168.50.92:5555",
            "name": "com.google.android.apps.nexuslauncher",
            "processId": 2743,
            "exited": false
        }
    ],
    "m_TagControl": {
        "m_Tags": [
            {
                "m_Name": "Filter by all listed tags",
                "m_Selected": false
            },
            {
                "m_Name": "No Filter",
                "m_Selected": false
            },
            {
                "m_Name": "",
                "m_Selected": false
            },
            {
                "m_Name": "Tag Control...",
                "m_Selected": false
            },
            {
                "m_Name": "",
                "m_Selected": false
            },
            {
                "m_Name": "Unity",
                "m_Selected": false
            },
            {
                "m_Name": "CRASH",
                "m_Selected": false
            },
            {
                "m_Name": "oidLogcatSampl",
                "m_Selected": true
            }
        ]
    },
    "m_MemoryViewerState": {
        "MemoryWindowHeight": 300.0,
        "MemoryWindowWidth": 170.0,
        "MemoryTypeEnabled": [
            true,
            true,
            true,
            true,
            true,
            true,
            true,
            true,
            true,
            true
        ],
        "MemoryGroup": 0,
        "Behavior": 1
    }
}
```


Juozas: 
* Can you check if logcat state is saved correctly - selected device, selected packages, tags, selected priority, as these items are now stored differently?

vincent-unity commented on com.unity.mobile.android-logcat/Editor/AndroidLogcatRuntime.cs:
```diff
@@ -7,94 +7,147 @@
+
+    internal class AndroidLogcatRuntime : AndroidLogcatRuntimeBase
+    {
+        private static readonly string kAndroidLogcatSettingsPath = Path.Combine("ProjectSettings", "AndroidLogcatSettings.asset");
```
Can we rename `kAndroidLogcatSettingsPath` to `kAndroidLogcatProjectSettingsPath`? As AndroidLogcatRuntimeBase contains both AndroidLogcatSettings & AndroidLogcatProjectSettings, this name makes me a little confused.

todi1856 replied:
yes, we can

todi1856 replied:
Went for kAndroidLogcatSettingsPath

todi1856 commented on com.unity.mobile.android-logcat/Editor/AndroidLogcatRuntime.cs:
```diff
@@ -7,94 +7,147 @@
-        AndroidLogcatDispatcher Dispatcher { get; }
+        protected AndroidLogcatDispatcher m_Dispatcher;
+        protected AndroidLogcatSettings m_Settings;
+        protected AndroidLogcatProjectSettings m_ProjectSettings;
```
the owner of the settings is runtime class, AndroidLogcatConsoleWindow is just a consumer. Looking ahead, if you want to get logcat without opening the AndroidLogcatConsoleWindow , it makes sense for settings to be created in a class which is not connected to UI

vincent-unity replied:
Ah, makes sense.

todi1856 commented on com.unity.mobile.android-logcat/Editor/AndroidLogcatConsoleWindow.cs:
```diff
@@ -140,88 +99,23 @@ public bool AutoSelectPackage
         }
 
-        private void OnEnable()
+        protected void OnInternalEnable(AndroidLogcatRuntimeBase runtime)
```
Fixed

todi1856 commented on com.unity.mobile.android-logcat/Editor/AndroidLogcatTagsControl.cs:
```diff
@@ -19,40 +19,44 @@ internal enum AndroidLogcatTagType
     internal class AndroidLogcatTagsControl
```
Fixed, but had to do a lot more refactoring since there was already Tags property inside

todi1856 commented on com.unity.mobile.android-logcat/Editor/AndroidLogcatRuntime.cs:
```diff
@@ -7,94 +7,147 @@
 namespace Unity.Android.Logcat
 {
-    internal interface IAndroidLogcatRuntime
+    internal abstract class AndroidLogcatRuntimeBase
```
all good then here

JuozasK commented:
Package selection doesn't get saved: 
Select a package, close the window, open it again, it says "No Filter"

Selected device isn't saved either:
Select Samsung Device, close the window, open it again, Huawei Nexus 6P is selected (the default out of the two)

Enabled/Disabled columns aren't saved:
Disable "Messages" column to not be shown, close and open the window - All the columns are visible

Regex field is saved
Tags and selected Tags are saved

Interestingly enough, all of the above are *saved* when shutting the editor off and restarting it.  Meaning that when you shut down the editor and restart it with the logcat window open, the package will be selected, the device will be the one selected previously and the Column selection will be preserved too. 

But then if I close the logcat window and open it again, now instead of Huawei Nexus 6P, the Samsung will be selected, which was selected when I had shut the Editor down.

todi1856 commented:
> Package selection doesn't get saved:
> Select a package, close the window, open it again, it says "No Filter"
> 
> Selected device isn't saved either:
> Select Samsung Device, close the window, open it again, Huawei Nexus 6P is selected (the default out of the two)
> 
> Enabled/Disabled columns aren't saved:
> Disable "Messages" column to not be shown, close and open the window - All the columns are visible
> 
> Regex field is saved
> Tags and selected Tags are saved
> 
> Interestingly enough, all of the above are _saved_ when shutting the editor off and restarting it. Meaning that when you shut down the editor and restart it with the logcat window open, the package will be selected, the device will be the one selected previously and the Column selection will be preserved too.
> 
> But then if I close the logcat window and open it again, now instead of Huawei Nexus 6P, the Samsung will be selected, which was selected when I had shut the Editor down.

Fixed:
*  Select a package, close the window, open it again, it says "No Filter"
*  Select Samsung Device, close the window, open it again, Huawei Nexus 6P is selected (the default out of the two)
*  Disable "Messages" column to not be shown, close and open the window - All the columns are visible (This one was not really a part of this PR, as that's the default behavior of alll Unity windows, but I went ahead and fixed that too)
